### ISSUE TITLE
Allow some SVG attributes to be displayed

### ISSUE THREAD

**Issue by ian-r-rose** [jupyterlab/jupyterlab#6010]
We currently strip SVGs from rendered HTML and Markdown. This can cause problems for some LaTeX renderers like KaTeX, which use SVG for a few elements. It would also be nice to allow basic SVGs in markdown, I might think.

According to @blois (thanks!) the following is the list of tags and attributes used by KaTeX, which we could consider adding to the allowed-list for our HTML sanitizer:

> the full list of SVG elements and attributes which are used by katex are:
> svg with attributes: width, height, viewBox, preserveAspectRatio and style
> line with attributes: x1, y1, x2, y2, stroke-width
> p with attribute: d

These would be added to the sanitizer here (as long as we decide they are okay for security reasons): https://github.com/jupyterlab/jupyterlab/blob/ae3bc129065705b96da79f2d16bef861f7a81c9b/packages/apputils/src/sanitizer.ts#L547-L702

**Comment by 97amarnathk** [comment]
Could I work on this?
It would be a good way to start contributing towards Jupyter :)

**Comment by blink1073** [comment]
Yes, please do!

**Comment by ggbhat** [comment]
I would like to work on this. @ian-r-rose  can you provide some KaTeX or LaTex example that renders svg output.
By white listing svg and line in senitizer I am able to display svg in markdown.

**Comment by ian-r-rose** [comment]
Hi @ggbhat I think one example of a LaTeX equation that fails to render with the KaTeX renderer due to SVG sanitizing is the `\vec` superscript, e.g.
```
$$
\vec{B}
$$
```

**Comment by ggbhat** [comment]
Thanks @ian-r-rose,   I don't see any svg element rendered from above superscript you have shared.  I have added svg and its attribute to sanitizer as you mentioned and trying to test it. But am not able to find proper superscript that verify the changes.

**Comment by ian-r-rose** [comment]
You can see how it's supposed to look by putting the same equation in the live renderer [here](https://katex.org/)

**Comment by jasongrout** [comment]
More context about svg in caja, etc., is available in links from this related issue for notebooks: https://github.com/jupyter/notebook/issues/4053

**Comment by vidartf** [comment]
[Here's a decent summary on some SVG security issues.](https://bjornjohansen.no/svg-in-wordpress) The main take-away should be that writing a sanitizer for SVG that actually works is hard. I'm not clear on whether the sanitizer we use claims to handle SVG sanitation properly?

**Comment by jasongrout** [comment]
Note that the PR for this is only opening up a small whitelist of attributes and elements.

**Comment by telamonian** [comment]
If we stick to just the basic graphical elements and attributes, the danger should be minimal. The problem come from the ability to embed all kinds of crazy xml tags inside of the svg tags. For example, you can put a `<script>` tag inside of an `<svg>` and execute arbitrary javascript that way.

In other words, allowing svg through will not (in theory) increase our attack surface area, since any tag that our sanitizer lets through inside of an svg, it will have already let through in other contexts anyway

**Comment by telamonian** [comment]
Having said that, I went back and checked the list of allowed tags to see if we were already letting through some of the more potentially dangerous inner svg elements:

https://github.com/jupyterlab/jupyterlab/blob/f17361c2d6ecf1b5ccae5b576266355150f09af2/packages/apputils/src/sanitizer.ts#L469-L573

There are a number of tags that can be used to execute code or link to another svg (`foreignObject`, `image`, `object`, `script`, `use`, [mdn docs here](https://developer.mozilla.org/en-US/docs/Web/SVG/Element)), and happily it looks like we don't whitelist any of them. Still, for future peace of mind, it would be nice if the sanitizer was able to conditionally whitelist tags based on their context

**Comment by ggbhat** [comment]
I agree with @telamonian already we are allowing some of the tags, allowing svg may not be an security risk.

**Comment by vidartf** [comment]
If you read the link I shared above, it details that the threat from user-content, inline SVGs are much broader than just script tags.

**Comment by vidartf** [comment]
E.g. here is a list of known exploits, that would all need to be ruled out (mostly for latest version of non-Opera browser, but still a decent size): http://html5sec.org/?svg.

E.g. the keylogger one is interesting: http://html5sec.org/?svg#132
